# 데이터 셋

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('/content/HR_Analytics.csv') # csv파일 로드
df.describe(include='all') # 데이터 전부 확인

Age Attrition BusinessTravel    DailyRate  \
count   1470.000000      1470           1470  1470.000000   
unique          NaN         2              3          NaN   
top             NaN        No  Travel_Rarely          NaN   
freq            NaN      1233           1043          NaN   
mean      36.923810       NaN            NaN   802.485714   
std        9.135373       NaN            NaN   403.509100   
min       18.000000       NaN            NaN   102.000000   
25%       30.000000       NaN            NaN   465.000000   
50%       36.000000       NaN            NaN   802.000000   
75%       43.000000       NaN            NaN  1157.000000   
max       60.000000       NaN            NaN  1499.000000   

                    Department  DistanceFromHome    Education EducationField  \
count                     1470       1470.000000  1470.000000           1470   
unique                       3               NaN          NaN              6   
top     Research & Development               NaN          NaN  Life Sciences   
freq                       961               NaN          NaN            606   
mean                       NaN          9.192517     2.912925            NaN   
std                        NaN          8.106864     1.024165            NaN   
min                        NaN          1.000000     1.000000            NaN   
25%                        NaN          2.000000     2.000000            NaN   
50%                        NaN          7.000000     3.000000            NaN   
75%                        NaN         14.000000     4.000000            NaN   
max                        NaN         29.000000     5.000000            NaN   

        EmployeeCount  EmployeeNumber  ...  RelationshipSatisfaction  \
count          1470.0     1470.000000  ...               1470.000000   
unique            NaN             NaN  ...                       NaN   
top               NaN             NaN  ...                       NaN   
freq              NaN             NaN  ...                       NaN   
mean              1.0     1024.865306  ...                  2.712245   
std               0.0      602.024335  ...                  1.081209   
min               1.0        1.000000  ...                  1.000000   
25%               1.0      491.250000  ...                  2.000000   
50%               1.0     1020.500000  ...                  3.000000   
75%               1.0     1555.750000  ...                  4.000000   
max               1.0     2068.000000  ...                  4.000000   

       StandardHours  StockOptionLevel  TotalWorkingYears  \
count         1470.0       1470.000000        1470.000000   
unique           NaN               NaN                NaN   
top              NaN               NaN                NaN   
freq             NaN               NaN                NaN   
mean            80.0          0.793878          11.279592   
std              0.0          0.852077           7.780782   
min             80.0          0.000000           0.000000   
25%             80.0          0.000000           6.000000   
50%             80.0          1.000000          10.000000   
75%             80.0          1.000000          15.000000   
max             80.0          3.000000          40.000000   

        TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
count             1470.000000     1470.000000     1470.000000   
unique                    NaN             NaN             NaN   
top                       NaN             NaN             NaN   
freq                      NaN             NaN             NaN   
mean                 2.799320        2.761224        7.008163   
std                  1.289271        0.706476        6.126525   
min                  0.000000        1.000000        0.000000   
25%                  2.000000        2.000000        3.000000   
50%                  3.000000        3.000000        5.000000   
75%                  3.000000        3.000000        9.000000   
max      

In [ ]:
df.columns # features 값 조회

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

## Data drop

In [2]:
df1 = df.drop(['BusinessTravel','Department','Education','EducationField','EmployeeCount','EmployeeNumber','JobInvolvement','JobLevel', 'JobRole', 'MaritalStatus',  'MonthlyRate', 'Over18', 'OverTime', 'PerformanceRating','StandardHours', 'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],axis=1) # 필요없는 feature 삭제
df2 = df1.loc[:,['Age','DailyRate','DistanceFromHome','EnvironmentSatisfaction','Gender','HourlyRate', 'JobSatisfaction',
        'MonthlyIncome', 'NumCompaniesWorked', 'PercentSalaryHike', 
       'RelationshipSatisfaction', 'StockOptionLevel',
         'WorkLifeBalance','Attrition']] # 종속변수값 재배치
df2.info() # 전처리 해야할 object 타입 확인

# X = df2.iloc[:,:-1].values # 독립변수 값들 OneHot인코딩을 위해 나누기
# cy = df2.iloc[:,-1:].values
# y = np.array([1 if x[0] == "Yes" else 0 for x in cy]) # 리스트 컴프리핸션으로 종속변수 YES값은 1로 NO값은 0으로 만들어 NUMPY배열로 변환환
# y
# # df2['Age'].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   DailyRate                 1470 non-null   int64 
 2   DistanceFromHome          1470 non-null   int64 
 3   EnvironmentSatisfaction   1470 non-null   int64 
 4   Gender                    1470 non-null   object
 5   HourlyRate                1470 non-null   int64 
 6   JobSatisfaction           1470 non-null   int64 
 7   MonthlyIncome             1470 non-null   int64 
 8   NumCompaniesWorked        1470 non-null   int64 
 9   PercentSalaryHike         1470 non-null   int64 
 10  RelationshipSatisfaction  1470 non-null   int64 
 11  StockOptionLevel          1470 non-null   int64 
 12  WorkLifeBalance           1470 non-null   int64 
 13  Attrition                 1470 non-null   object
dtypes: int64(12), object(2)


In [ ]:
df2['WorkLifeBalance'].describe()
df2['MonthlyIncome'].describe()

count     1470.000000
mean      6502.931293
std       4707.956783
min       1009.000000
25%       2911.000000
50%       4919.000000
75%       8379.000000
max      19999.000000
Name: MonthlyIncome, dtype: float64

## OneHotEnconding - Getdummies

In [3]:
# from sklearn.compose import ColumnTransformer # 원 핫 인코드
# from sklearn.preprocessing import OneHotEncoder
# ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(drop='first'), [4])], remainder='passthrough')
# ct
# X = ct.fit_transform(X)
# X

X = pd.get_dummies(df2, columns=['Gender'], drop_first=True)
X = X.drop('Attrition', axis=1)
y = df2['Attrition'].apply(lambda x : 1 if x == "Yes" else 0)

# cy = df2.iloc[:,-1:].values
# y = np.array([1 if x[0] == "Yes" else 0 for x in cy]) # 리스트 컴프리핸션으로 종속변수 YES값은 1로 NO값은 0으로 만들어 NUMPY배열로 변환환


## 테스트셋 분리

In [4]:
from sklearn.model_selection import train_test_split # 테스트셋, 트레인셋 분류
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=20)

In [ ]:
X, len(X)

(array([[0.0, 41, 1102, ..., 1, 0, 1],
        [1.0, 49, 279, ..., 4, 1, 3],
        [1.0, 37, 1373, ..., 2, 0, 3],
        ...,
        [1.0, 27, 155, ..., 2, 1, 3],
        [1.0, 49, 1023, ..., 4, 0, 2],
        [1.0, 34, 628, ..., 1, 0, 4]], dtype=object), 1470)

In [ ]:
X_train, len(X_train) # 훈련 세트 X, 개수

(array([[1.0, 27, 1115, ..., 4, 0, 3],
        [1.0, 28, 1158, ..., 4, 1, 2],
        [1.0, 35, 1370, ..., 2, 1, 3],
        ...,
        [1.0, 47, 666, ..., 4, 1, 2],
        [1.0, 24, 691, ..., 2, 2, 3],
        [0.0, 27, 1337, ..., 1, 0, 3]], dtype=object), 1323)

In [ ]:
X_test, len(X_test) # 테스트 세트 X, 개수

(array([[1.0, 38, 833, ..., 3, 1, 3],
        [1.0, 31, 1274, ..., 4, 1, 4],
        [0.0, 33, 553, ..., 2, 2, 3],
        ...,
        [1.0, 28, 1476, ..., 3, 0, 3],
        [0.0, 34, 1354, ..., 4, 0, 3],
        [1.0, 38, 1035, ..., 3, 0, 3]], dtype=object), 147)

In [ ]:
y, len(y) # 전체 데이터 y

(array([1, 0, 1, ..., 0, 0, 0]), 1470)

In [ ]:
y_train, len(y_train)

(array([0, 0, 0, ..., 1, 0, 1]), 1323)

In [ ]:
y_test, len(y_test)

(array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
        1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
        0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
        0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), 147)

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   DailyRate                 1470 non-null   int64 
 2   DistanceFromHome          1470 non-null   int64 
 3   EnvironmentSatisfaction   1470 non-null   int64 
 4   Gender                    1470 non-null   object
 5   HourlyRate                1470 non-null   int64 
 6   JobSatisfaction           1470 non-null   int64 
 7   MonthlyIncome             1470 non-null   int64 
 8   NumCompaniesWorked        1470 non-null   int64 
 9   PercentSalaryHike         1470 non-null   int64 
 10  RelationshipSatisfaction  1470 non-null   int64 
 11  StockOptionLevel          1470 non-null   int64 
 12  WorkLifeBalance           1470 non-null   int64 
 13  Attrition                 1470 non-null   object
dtypes: int64(12), object(2)


# 랜덤 포레스트


## importance check

In [11]:
# 랜덤포레스트 분류모델
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

clf = RandomForestClassifier(n_estimators=15, random_state=42)
clf.fit(X_train, y_train)
train_pred = clf.predict(X_train)
test_pred = clf.predict(X_test)

# sample_data = [49,279,8,3,1,61,2,5130,1,23,4,1,3]
# np_arr = np.array(sample_data)
# np_arr = np_arr.reshape(1,-1)
# leave = clf.predict(np_arr)

# 정확도를 계산하여 모델의 성능을 평가합니다.
accuracy = accuracy_score(y_test, test_pred)
print('Accuracy:', accuracy)
print(f'Train-set : {clf.score(X_train, y_train)}')
print(f'Test-set : {clf.score(X_test, y_test)}')

for i, importance in enumerate(clf.feature_importances_):
    print("Feature ", i, X_train.columns[i], " importance: ", importance)


# !pip install mlxtend -q

# from google.colab import files
# import joblib # pkl 세이브파일을 만들어주는 라이브러리

# joblib.dump(clf, 'RandomForest_model.pkl')
# files.download('RandomForest_model.pkl') # 다운로드
# model_from_joblib = joblib.load('RandomForest_model.pkl')


# # df2.info() # 전처리 해야할 object 타입 확인

Accuracy: 0.9727891156462585
Train-set : 0.9957482993197279
Test-set : 0.9727891156462585
Feature  0 Age  importance:  0.11046911139770632
Feature  1 DailyRate  importance:  0.11681263599762827
Feature  2 DistanceFromHome  importance:  0.07948335154370835
Feature  3 EnvironmentSatisfaction  importance:  0.044400745646506985
Feature  4 HourlyRate  importance:  0.09687409476314926
Feature  5 JobSatisfaction  importance:  0.04271794051595066
Feature  6 MonthlyIncome  importance:  0.1937241900375066
Feature  7 NumCompaniesWorked  importance:  0.06809700914809598
Feature  8 PercentSalaryHike  importance:  0.08906093361068268
Feature  9 RelationshipSatisfaction  importance:  0.051079145183655424
Feature  10 StockOptionLevel  importance:  0.05728353426269031
Feature  11 WorkLifeBalance  importance:  0.035023011982152234
Feature  12 Gender_Male  importance:  0.014974295910566931


# Grid서치 하이퍼파라미터 튜닝 + *랜덤포레스트*

In [ ]:
# 위의 랜덤포레스트 분류형 모델 코드를 GridSearchCV 하이퍼파라미터 튜닝
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

param_grid = {
    'n_estimators': [50, 100, 150],
    'max_features': ['sqrt', 'log2'],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 4, 6],
    'min_samples_leaf': [1, 2, 4]
}

rf = RandomForestClassifier(random_state=42)

grid_search = GridSearchCV(rf, param_grid=param_grid, cv=5, n_jobs=-1)
grid_search.fit(X_train, y_train)

print("Best parameters:", grid_search.best_params_)
print("Best score:", grid_search.best_score_)

# 최적의 파라미터로 학습된 모델로 테스트 데이터 예측
y_pred = grid_search.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)


Best parameters: {'max_depth': None, 'max_features': 'sqrt', 'min_samples_leaf': 2, 'min_samples_split': 6, 'n_estimators': 50}
Best score: 0.8442967409948542
Accuracy: 0.8639455782312925


In [ ]:
df_col = df2.copy()
df_col.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   DailyRate                 1470 non-null   int64 
 2   DistanceFromHome          1470 non-null   int64 
 3   EnvironmentSatisfaction   1470 non-null   int64 
 4   Gender                    1470 non-null   object
 5   HourlyRate                1470 non-null   int64 
 6   JobSatisfaction           1470 non-null   int64 
 7   MonthlyIncome             1470 non-null   int64 
 8   NumCompaniesWorked        1470 non-null   int64 
 9   PercentSalaryHike         1470 non-null   int64 
 10  RelationshipSatisfaction  1470 non-null   int64 
 11  StockOptionLevel          1470 non-null   int64 
 12  WorkLifeBalance           1470 non-null   int64 
 13  Attrition                 1470 non-null   object
dtypes: int64(12), object(2)


In [ ]:
df2

Age  DailyRate  DistanceFromHome  EnvironmentSatisfaction  Gender  \
0      41       1102                 1                        2  Female   
1      49        279                 8                        3    Male   
2      37       1373                 2                        4    Male   
3      33       1392                 3                        4  Female   
4      27        591                 2                        1    Male   
...   ...        ...               ...                      ...     ...   
1465   36        884                23                        3    Male   
1466   39        613                 6                        4    Male   
1467   27        155                 4                        2    Male   
1468   49       1023                 2                        4    Male   
1469   34        628                 8                        2    Male   

      HourlyRate  JobSatisfaction  MonthlyIncome  NumCompaniesWorked  \
0             94                4           5993                   8   
1             61                2           5130                   1   
2             92                3           2090                   6   
3             56                3           2909                   1   
4             40                2           3468                   9   
...          ...              ...            ...                 ...   
1465          41                4           2571                   4   
1466          42                1           9991                   4   
1467          87                2           6142                   1   
1468          63                2           5390                   2   
1469          82                3           4404                   2   

      PercentSalaryHike  RelationshipSatisfaction  StockOptionLevel  \
0                    11                         1                 0   
1                    23                         4                 1   
2                    15                         2                 0   
3                    11                         3                 0   
4                    12                         4                 1   
...                 ...                       ...               ...   
1465                 17                         3                 1   
1466                 15                         1                 1   
1467                 20                         2                 1   
1468                 14                         4                 0   
1469                 12                         1                 0   

      WorkLifeBalance Attrition  
0                   1       Yes  
1                   3        No  
2                   3       Yes  
3                   3        No  
4                   3        No  
...               ...       ...  
1465                3        No  
1466                3        No  
1467                3        No  
1468                2        No  
1469                4        No  

[1470 rows x 14 columns]

# LightGBM

## Importance Check

In [ ]:
# LightGBM 모델 사용
import lightgbm as lgb
from sklearn.metrics import accuracy_score
lgb_model = lgb.LGBMClassifier(n_estimators=15, random_state=42,  num_iterations= 50) 
# n_estimators 15개의 결정트리 생성
# num_iterations 반복횟수 50회
lgb_model.fit(X_train, y_train)
train_pred = lgb_model.predict(X_train)
test_pred = lgb_model.predict(X_test)



print(f'Train-set : {clf.score(X_train, y_train)}')
print(f'Test-set : {clf.score(X_test, y_test)}')

accuracy = accuracy_score(y_test, test_pred)
print('Accuracy:', accuracy)

for i, importance in enumerate(lgb_model.feature_importances_):
    print("Feature ", i, X_train.columns[i], " importance: ", importance)
# !pip install mlxtend -q

# from google.colab import files
# import joblib # pkl 세이브파일을 만들어주는 라이브러리

# joblib.dump(lgb_model, 'LightGBM_model.pkl')

# files.download('LightGBM_model.pkl') # 다운로드
# model_from_joblib = joblib.load('LightGBM_model.pkl')
# model_from_joblib


/usr/local/lib/python3.9/dist-packages/lightgbm/engine.py:177: UserWarning: Found `num_iterations` in params. Will use it instead of argument
  _log_warning(f"Found `{alias}` in params. Will use it instead of argument")


Train-set : 0.9969765684051398
Test-set : 0.891156462585034
Accuracy: 0.8503401360544217


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

LGBMClassifier(n_estimators=15, num_iterations=50, random_state=42)

# Grid서치 하이퍼파라미터 튜닝 + LightGBM

In [ ]:
# 위의 LightGBM 모델 코드를 GridSearchCV 하이퍼파라미터 튜닝
from sklearn.model_selection import GridSearchCV

lgb_model = lgb.LGBMClassifier(n_estimators=100, random_state=42)

# 하이퍼파라미터 후보군 설정
param_grid = {
    'max_depth': [3, 5, 7],
    'learning_rate': [0.05, 0.1, 0.15]
}

# GridSearchCV를 이용한 하이퍼파라미터 튜닝
grid_search = GridSearchCV(
    estimator=lgb_model, 
    param_grid=param_grid,
    cv=5, # 교차 검증 횟수
    n_jobs=-1 # 모든 CPU 코어 사용
)

grid_search.fit(X_train, y_train)

# 최적의 하이퍼파라미터 조합
best_params = grid_search.best_params_
print(best_params)

# 최적의 모델
best_model = grid_search.best_estimator_

# 테스트 세트에 대한 정확도
accuracy = accuracy_score(y_test, best_model.predict(X_test))
print('Accuracy:', accuracy)

# !pip install mlxtend -q

# from google.colab import files
# import joblib # pkl 세이브파일을 만들어주는 라이브러리

# joblib.dump(best_model, 'Tuned_LightGBM_model.pkl')

# files.download('Tuned_LightGBM_model.pkl') # 다운로드
# model_from_joblib = joblib.load('Tuned_LightGBM_model.pkl')
# model_from_joblib

{'learning_rate': 0.05, 'max_depth': 3}
Accuracy: 0.8707482993197279


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

LGBMClassifier(learning_rate=0.05, max_depth=3, random_state=42)

# XGBoost

## Importance Check

In [10]:
# XGBoost 모델 사용
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV

# 훈련 및 검증 데이터 분할
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, random_state=42)
xgb_model = XGBClassifier(n_estimators=200, learning_rate=0.1, max_depth=3, min_child_weight=3, colsample_bytree=1)
xgb_model.fit(X_train, y_train, early_stopping_rounds=200, eval_metric='auc', eval_set=[(X_valid, y_valid)])
y_pred = xgb_model.predict(X_valid)

# 정확도 계산
acc = accuracy_score(y_valid, y_pred)
acc
dict = {}
for i, importance in enumerate(xgb_model.feature_importances_):
    print("Feature ", i, X_train.columns[i], " importance: ", importance)
    dict[X_train.columns[i]] = importance
idf = pd.DataFrame([dict])
idf

# !pip install mlxtend -q

# from google.colab import files
# import joblib # pkl 세이브파일을 만들어주는 라이브러리

# joblib.dump(xgb_model, 'XGBoost.pkl')

# files.download('XGBoost.pkl') # 다운로드
# model_from_joblib = joblib.load('XGBoost.pkl')
# model_from_joblib

[0]	validation_0-auc:0.58326
[1]	validation_0-auc:0.58924
[2]	validation_0-auc:0.59492
[3]	validation_0-auc:0.59115
[4]	validation_0-auc:0.60166
[5]	validation_0-auc:0.60548
[6]	validation_0-auc:0.60322
[7]	validation_0-auc:0.60784
[8]	validation_0-auc:0.63007
[9]	validation_0-auc:0.62081
[10]	validation_0-auc:0.62217
[11]	validation_0-auc:0.63238
[12]	validation_0-auc:0.63444
[13]	validation_0-auc:0.64570
[14]	validation_0-auc:0.65500
[15]	validation_0-auc:0.65601
[16]	validation_0-auc:0.65455
[17]	validation_0-auc:0.65344
[18]	validation_0-auc:0.65631
[19]	validation_0-auc:0.65676
[20]	validation_0-auc:0.66048
[21]	validation_0-auc:0.66440
[22]	validation_0-auc:0.67029
[23]	validation_0-auc:0.67793
[24]	validation_0-auc:0.68160
[25]	validation_0-auc:0.68059
[26]	validation_0-auc:0.68019
[27]	validation_0-auc:0.68200
[28]	validation_0-auc:0.68632
[29]	validation_0-auc:0.69075
[30]	validation_0-auc:0.69110
[31]	validation_0-auc:0.69537


/usr/local/lib/python3.9/dist-packages/xgboost/sklearn.py:835: UserWarning: `eval_metric` in `fit` method is deprecated for better compatibility with scikit-learn, use `eval_metric` in constructor or`set_params` instead.
  warnings.warn(
/usr/local/lib/python3.9/dist-packages/xgboost/sklearn.py:835: UserWarning: `early_stopping_rounds` in `fit` method is deprecated for better compatibility with scikit-learn, use `early_stopping_rounds` in constructor or`set_params` instead.
  warnings.warn(


[32]	validation_0-auc:0.69789
[33]	validation_0-auc:0.69920
[34]	validation_0-auc:0.70241
[35]	validation_0-auc:0.70548
[36]	validation_0-auc:0.70327
[37]	validation_0-auc:0.70005
[38]	validation_0-auc:0.70146
[39]	validation_0-auc:0.70669
[40]	validation_0-auc:0.70689
[41]	validation_0-auc:0.71403
[42]	validation_0-auc:0.71835
[43]	validation_0-auc:0.72197
[44]	validation_0-auc:0.71664
[45]	validation_0-auc:0.71926
[46]	validation_0-auc:0.71694
[47]	validation_0-auc:0.71916
[48]	validation_0-auc:0.71905
[49]	validation_0-auc:0.71724
[50]	validation_0-auc:0.71885
[51]	validation_0-auc:0.72278
[52]	validation_0-auc:0.72408
[53]	validation_0-auc:0.72167
[54]	validation_0-auc:0.72328
[55]	validation_0-auc:0.72609
[56]	validation_0-auc:0.72579
[57]	validation_0-auc:0.72760
[58]	validation_0-auc:0.72760
[59]	validation_0-auc:0.72831
[60]	validation_0-auc:0.73132
[61]	validation_0-auc:0.73122
[62]	validation_0-auc:0.72891
[63]	validation_0-auc:0.73203
[64]	validation_0-auc:0.73353
[65]	valid

Age  DailyRate  DistanceFromHome  EnvironmentSatisfaction  HourlyRate  \
0  0.078358    0.05182          0.066009                 0.102123    0.052345   

   JobSatisfaction  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
0         0.079136        0.08982            0.085802           0.055501   

   RelationshipSatisfaction  StockOptionLevel  WorkLifeBalance  Gender_Male  
0                  0.076189          0.126147         0.058167     0.078584

# Grid서치 하이퍼파라미터 튜닝 + XGBoost

In [ ]:
# 위의 XGboost 모델 코드를 GridSearchCV 하이퍼파라미터 튜닝
import xgboost as xgb
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score
from xgboost import XGBClassifier

# 훈련 및 검증 데이터 분할
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, random_state=42)

# XGBClassifier 객체 생성
xgb_model = XGBClassifier()

# GridSearchCV를 위한 하이퍼 파라미터 설정
param_grid = {
    'max_depth': [3, 5, 7],
    'learning_rate': [0.1, 0.01, 0.001],
    'n_estimators': [100, 200, 300],
    'min_child_weight': [1, 3, 5],
    'colsample_bytree': [0.7, 0.8, 0.9, 1.0]
}

# GridSearchCV 객체 생성
grid = GridSearchCV(xgb_model, param_grid, cv=3, n_jobs=-1, scoring='accuracy', verbose=3)

# GridSearchCV로 모델 훈련
grid.fit(X_train, y_train)

# 최적의 하이퍼 파라미터 출력
print(f'Best parameter: {grid.best_params_}')

# 최적의 모델로 예측 및 검증
y_pred = grid.predict(X_valid)
acc = accuracy_score(y_valid, y_pred)
print(f'Accuracy: {acc}')


# !pip install mlxtend -q

# from google.colab import files
# import joblib # pkl 세이브파일을 만들어주는 라이브러리

# joblib.dump(grid, 'Tuned_XGBoost.pkl')

# files.download('Tuned_XGBoost.pkl') # 다운로드
# model_from_joblib = joblib.load('Tuned_XGBoost.pkl')
# model_from_joblib


Fitting 3 folds for each of 324 candidates, totalling 972 fits
Best parameter: {'colsample_bytree': 0.7, 'learning_rate': 0.001, 'max_depth': 3, 'min_child_weight': 5, 'n_estimators': 200}
Accuracy: 0.8673469387755102


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

GridSearchCV(cv=3,
             estimator=XGBClassifier(base_score=None, booster=None,
                                     callbacks=None, colsample_bylevel=None,
                                     colsample_bynode=None,
                                     colsample_bytree=None,
                                     early_stopping_rounds=None,
                                     enable_categorical=False, eval_metric=None,
                                     feature_types=None, gamma=None,
                                     gpu_id=None, grow_policy=None,
                                     importance_type=None,
                                     interaction_constraints=None,
                                     learning_rate=None,...
                                     max_leaves=None, min_child_weight=None,
                                     missing=nan, monotone_constraints=None,
                                     n_estimators=100, n_jobs=None,
                                     num_parallel_tree=None, predictor=None,
                                     random_state=None, ...),
             n_jobs=-1,
             param_grid={'colsample_bytree': [0.7, 0.8, 0.9, 1.0],
                         'learning_rate': [0.1, 0.01, 0.001],
                         'max_depth': [3, 5, 7], 'min_child_weight': [1, 3, 5],
                         'n_estimators': [100, 200, 300]},
             scoring='accuracy', verbose=3)

# 랜덤포레스트 **회귀형** 모델 GridSearchCV 하이퍼파라미터 튜닝 전
* Mean squared error: 0.023083900226757367
+ Train score: 0.8976170484669462
+ Test score: 0.8036141732283465

# 랜덤포레스트 **회귀형** 모델 GridSearchCV 하이퍼파라미터 튜닝 후
+ Best parameters: {'max_depth': 10, 'max_features': 'sqrt', 'n_estimators': 150}
* Best score: 0.21560769029234592
* Mean squared error: 0.028595682399876726
* Train score: 0.8105706921316806
* Test score: 0.7567227948901826

# 랜덤포레스트 **분류형** 모델 GridSearchCV 하이퍼파라미터 튜닝 전
* Accuracy: 0.9727891156462585

# 랜덤포레스트 **분류형** 모델 GridSearchCV 하이퍼파라미터 튜닝 후
* Best parameters: {'max_depth': 10, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 4, 'n_estimators': 50}
* Best score: 0.8563000360620266
* Accuracy: 0.9591836734693877

# LightGBM 모델 GridSearchCV 하이퍼파라미터 튜닝 전
* Accuracy: 0.9047619047619048

# LightGBM 모델 GridSearchCV 하이퍼파라미터 튜닝 후
* {'learning_rate': 0.05, 'max_depth': 3}
* Accuracy: 0.9319727891156463

# XGBoost 모델 GridSearchCV 하이퍼파라미터 튜닝 전
* Accuracy: 0.8877551020408163

# XGBoost 모델 GridSearchCV 하이퍼파라미터 튜닝 후
* Best parameter: {'colsample_bytree': 0.7, 'learning_rate': 0.1, 'max_depth': 5, 'min_child_weight': 5, 'n_estimators': 200}
* Accuracy: 0.891156462585034